Allow to send headers to proxied documents

marcospri opened the pull request:
For:

- https://github.com/hypothesis/lms/issues/4768

## Testing

Follow the instructions over: https://github.com/hypothesis/lms/pull/4792

marcospri committed:
Secret AES encryption helper

Allows to encrypt dictionaries using AES and encode both the cipher text
and IV in a base64 encoded string.

marcospri committed:
Take headers in url_for

Take an optional `headers` parameters which wild be sent AES encoded
over the wire to Via.

marcospri commented on src/h_vialib/client.py:
```diff
@@ -45,6 +45,7 @@ def __init__(self, secret, service_url=None, html_service_url=None):
         :param service_url: Location of the via server
         :param html_service_url: Location of the Via HTML presenter
         """
+        self._secure_secrets = SecureSecrets(secret.encode("utf-8"))
```
No need to double guess inside `SecureSecrets`, here it's a `str`, conver it to `bytes`.

seanh replied:
I guess we could alternatively make it LMS and Via/Via HTML's jobs to pass in a byte string to h_vialib in the first place. They could either encode it or perhaps they could use [environb](https://docs.python.org/3/library/os.html#os.environb) to write it as a byte string directly from the environ. But this seems fine

marcospri replied:
We use `secret: str`, not a new parameter, elsewhere, we need it to be bytes here.

marcospri commented on src/h_vialib/client.py:
```diff
@@ -56,7 +57,10 @@ def __init__(self, secret, service_url=None, html_service_url=None):
         }
 
-    def url_for(self, url, content_type=None, options=None, blocked_for=None):
+    # pylint:disable=too-many-arguments
```
I reckon this might keep growing and have a similar interface to requests.request.

Not a problem IMO.

seanh replied:
:+1:

marcospri commented on src/h_vialib/client.py:
```diff
@@ -74,6 +80,9 @@ def url_for(self, url, content_type=None, options=None, blocked_for=None):
             query.update(options)
 
+        if headers:
+            query["via.secret.headers"] = self._secure_secrets.encrypt_dict(headers)
```
`_secure_secrets` interface seems to be built only for this case but... it has.

marcospri commented on src/h_vialib/secure/secrets.py:
```diff
@@ -0,0 +1,49 @@
+    def __init__(self, secret: bytes):
+        self._secret = pad(secret, 16)
+
+    def encrypt_dict(self, payload: dict) -> str:
```
Encrypting the dict (converting it to json etc) seems useful to have here rather that just expose low lever primitives.

marcospri commented on src/h_vialib/secure/secrets.py:
```diff
@@ -0,0 +1,49 @@
+        dict_json = json.dumps(payload).encode("utf-8")
+        aes_iv, encrypted_json = self._encrypt(dict_json)
+
+        return json.dumps(
```
Maybe this whole thing should be base64 too?

marcospri commented on src/h_vialib/secure/secrets.py:
```diff
@@ -0,0 +1,49 @@
+
+        return json.loads(self._decrypt(aes_iv, cipher))
+
+    def _decrypt(self, aes_iv, encrypted) -> bytes:
```
Making these private. We might need them in the future but as this is a library we want to keep the interface under control.

seanh replied:
So the existing `SecureToken` and `SecureURL` classes aren't appropriate for this use-case? I guess not because it looks like `SecureURL` only signs the URL doesn't encrypt it.

I wonder if we should rename some of these classes, to clarify what they do and don't do? For example perhaps `SecureURL` should be `SignedURL` if it just signs and doesn't encrypt. I think `SecureSecrets` encrypts a dict but doesn't sign it, so perhaps `EncryptedDict`?

I think we should probably stop using the words "secret" and "secure" because it's not clear exactly what they mean, and maybe instead say "signed" and "encrypted".

marcospri replied:
EncryptedDict is a better name :+1:.

marcospri replied:
I wrote this as `EncryptedDict` but from the callers perspetive it looked like it was as `dict` subclass supporting encryption. It didn't feel right.

I renamed the class to a more explicit `Encryption` and keep the method names.

marcospri commented on tests/unit/h_vialib/secure/secrets_test.py:
```diff
@@ -0,0 +1,70 @@
+
+
+class TestSecureSecrets:
+    def test_encrypt_dict_round_trip(self, secure_secrets):
```
The whole process has a lot of conversions back and forwards so this tests is very useful.

The rest are there to jsut assert that we are in fact calling AES.

seanh commented on .cookiecutter/includes/setuptools/install_requires:
```diff
@@ -1,2 +1,3 @@
+pycryptodomex
```
Any notes on why pycryptodomex?

marcospri replied:
This is the PyCrypto replacement we chose for LMS, make sense to use the same:

https://github.com/hypothesis/lms/commit/c22cc56c8fb842d18e92d348901d8876d57358b5
https://github.com/hypothesis/lms/commit/6db9c65f69e3280f29f97ec4d686014df58baa6a

seanh commented on src/h_vialib/client.py:
```diff
@@ -56,7 +57,10 @@ def __init__(self, secret, service_url=None, html_service_url=None):
 
-    def url_for(self, url, content_type=None, options=None, blocked_for=None):
+    # pylint:disable=too-many-arguments
+    def url_for(
+        self, url, content_type=None, options=None, blocked_for=None, headers=None
+    ):
```
`headers` are only supported by certain Via endpoints, right? The ones that proxy in Python, not the ones that proxy in NGINX. Are there some `url`'s where `url_for(url, headers=my_headers)` will return a URL containing headers that Via will actually ignore because it's going to proxy that URL through NGINX?

marcospri replied:
> will return a URL containing headers that Via will actually ignore because it's going to proxy that URL through NGINX?

Yes, all HTML ones and pdfs not proxied by python (so all of them except one drive and d2l ones).

Applying the headers to everything will be a big tasks (not sure even if possible without writing nginx lua code).

Maybe we could just explain the limitation on the method's docstring?

marcospri replied:
Added a comment about this. Not ideal but naming this `headers_for_python_proxing` is not better.

seanh replied:
Maybe we could make Via throw some sort of "not implemented" error if any `headers` are passed to it and it would have proxied the request through NGINX rather than through Python?

seanh reviewed (commented):
Posting some initial questions. I want to come back and take a closer look at the crypto code in here when I have more time, just to double-check...

marcospri committed:
Rename SecureSecrets to Encryption

marcospri committed:
Switch over JWEs

Instead of our own mix of base64,AES,iv... just use JWE for the same
functionality.

marcospri commented on .cookiecutter/includes/setuptools/install_requires:
```diff
@@ -1,2 +1,2 @@
-pyjwt
+python-jose[cryptography]
```
python jose supports JWE.

We already use it on LMS for it's support for JWKs and the plan is to settle in just this one library for JWT related needs.

marcospri commented on src/h_vialib/secure/encryption.py:
```diff
@@ -0,0 +1,26 @@
+    def __init__(self, secret: bytes):
+        self._secret = secret.ljust(32)[:32]
+
+    def encrypt_dict(self, payload: dict) -> str:
```
Using JWE simplified this a lot.

marcospri commented on src/h_vialib/secure/token.py:
```diff
@@ -47,9 +47,7 @@ def verify(self, token):
             raise InvalidToken("Expired secure token") from err
-        except DecodeError as err:
-            raise InvalidToken("Malformed secure token") from err
+        except JWTError as err:
```
python-jose doesn't expose these differently. In any case we are wrapping them in the same `InvalidToken` exception in our end.

marcospri commented on tests/unit/h_vialib/secure/token_test.py:
```diff
@@ -14,6 +15,32 @@ def decode_token(token_string):
 
 
 class TestSecureToken:
+    @freeze_time("2022-12-22")
```
Using freezegun so the output is always the same.

marcospri commented on tests/unit/h_vialib/secure/token_test.py:
```diff
@@ -14,6 +15,32 @@ def decode_token(token_string):
+            (
+                {"b": 5},
+                timedelta(seconds=10),
+                "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJiIjo1LCJleHAiOjE2NzE2NjcyMTB9.Gf1e9jLnn57iGGU4c7nit6zIZ6LWNIQTHiOzk9cLJ9c",
```
These test make the jwt library replacement feel safer.

seanh reviewed (approved):
This all looks good to me. Made a comment below about potentially wanting to make Via (not `h-vialib`) error if `headers` are sent for an NGINX or Via HTML-proxied URL where `headers` aren't supported
